app/models/request.rb

```diff
@@ -63,31 +53,55 @@ class Request < ActiveRecord::Base
     end
   end
 
+  def cancel
+    if self.status == "valet_pick_up" || "request_pick_up"
+      self.cancel_pick_up!
+    elsif self.status == "request_drop_off" || "valet_drop_off"
+      self.cancel_drop_off!
+    end
+  end
+
   def remove_valet_pick_up
     unless self.valet_pick_up.nil?
-      self.update(valet_pick_up:  nil)
+      self.update(valet_pick_up:  nil,
+      car_pick_up_time: nil)
     end
   end
 
   def remove_valet_drop_off
     unless self.valet_drop_off.nil?
-      self.update(valet_drop_off:  nil)
+      self.update(valet_drop_off:  nil,
+        request_drop_off_time: nil)
     end
   end
 
   def generate_auth_code(drop_off = false)
     auth_code = '%04d' % rand(10 ** 4)
-    if drop_off
-      self.update(auth_code_drop_off: auth_code)
-    else
-      self.update(auth_code_pick_up: auth_code)
-    end
+    drop_off ? self.update(auth_code_drop_off: auth_code) : self.update(auth_code_pick_up: auth_code)
   end
 
   def find_nearest_parking
     self.update(parking_location: ParkingLot.near(self.source_location, 10, units: :km).first)
   end
 
+  def auth_code_check?(auth_code, dropoff = false)
+    dropoff ? self.auth_code_drop_off == auth_code : self.auth_code_pick_up == auth_code
```

Comment:
It checks to see which field it needs to add to.  If you send in true as a parameter it will put it into drop_off field instead so I can use this function for both the auth_code field for drop off and the auth_code field for pickup